### PR TITLE
add stub for LimitIterator

### PR DESCRIPTION
**Pull request by Khartir**

### CODE REVIEW

File: stubs/iterable.stub
Comment (ondrejmirtes) [inline stubs/iterable.stub:288]: Needs to be added here: https://github.com/phpstan/phpstan-src/blob/2b3d28244afa01be2dc7df71e6879bc5fd99b1b5/conf/config.neon#L25-L29

Then you probably won't get that extra error.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by staabm** [comment]
after this PR has been merged, you might have fun adding even more Iterators (one by one with a separate PR)

see https://github.com/phpstan/phpstan-src/pull/1009 for inspiration

note https://github.com/phpstan/phpstan-src/pull/1009#issuecomment-1075160461 though

File: stubs/iterable.stub
Comment (Khartir) [reply stubs/iterable.stub:288]: Done, but the error remains. I assume this is because  the generic is incomplete instead of missing?

File: tests/PHPStan/Analyser/data/iterator-iterator.php
Comment (staabm) [inline tests/PHPStan/Analyser/data/iterator-iterator.php:26]: shouldn't this return `\ArrayIterator<int, string>`?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
We should be compatible with Psalm's stub that has only two `@template`: https://github.com/vimeo/psalm/blob/36d5a2a83c1d7b3e2aca1cfe9e9dcc916619e962/stubs/CoreGenericIterators.phpstub#L552-L577

If you disagree with it, we first need to open a discussion with Psalm's maintainers.

**Comment by Khartir** [comment]
> We should be compatible with Psalm's stub

I looked into this and as far as I can tell, yes there needs to be a discussion. Both [Psalm](https://github.com/vimeo/psalm/blob/36d5a2a83c1d7b3e2aca1cfe9e9dcc916619e962/stubs/CoreGenericIterators.phpstub#L105-L107) and [PHPStan](https://github.com/phpstan/phpstan-src/blob/eeedd1cbcf45a6c248ed92621ab7faff4172df8d/stubs/iterable.stub#L181-L183) define `IteratorIterator` with 3 `@template`.
However Psalm then only has two `@template` for all extensions of `IteratorIterator`. As far as I can tell this is not possible in PHPStan.
For this reason, there are already some differences: `FilterIterator`, `CallbackFilterIterator` and `RecursiveCallbackFilterIterator` are all defined with 3 `@template` in PHPStan, while they have only two in Psalm.

Furthermore, as @staabm pointed out, `OuterIterator` should probably be changed to also have 3 `@template`, so that the type of `getInnerIterator` is correct:
```php
/**
 * @template-covariant TKey
 * @template-covariant TValue
 * @template TIterator as Iterator<TKey, TValue>
 *
 * @template-extends Iterator<TKey, TValue>
 */
interface OuterIterator extends Iterator {
    /**
     * @return TIterator
     */
    public function getInnerIterator();
}
```
This is also missing in Psalm.

**Comment by Khartir** [comment]
On further testing I realized that changing `OuterIterator` leads to similar issues as discussed in https://github.com/phpstan/phpstan/issues/6829, because it assumes that the `Iterator` returned from `getInnerIterator` is the same as the one passed to the constructor in `IteratorIterator` (and `RecursiveIteratorIterator` which is currently missing `implements OuterIterator`). This wouldn't even work, because both constructors accept `Traversable`, not `Iterator`.

**Comment by ondrejmirtes** [comment]
Hi, I'm cleaning up old and stale PRs. Please send a new PR if you're still interested, thanks.

**Pull request closed without merging**
